### ISSUE TITLE
Improve diagnostic for mutable containers involving variance

### ISSUE THREAD

**Issue by honnorat** [astral-sh/ty#2976]
### Summary

With new version 0.0.21, an `invalid-argument-type` error pops when a `dict[str, str]` is passed as a variable (not as a value) to a function expecting a `dict[str, str | None]`.

See: https://play.ty.dev/822e13ba-aa8b-4314-b6b6-0aa3b825920a

```python
## File: dict_str.py
def func(attrs: dict[str, str | None]) -> None:
    pass

func({"key": "value"})

d = {"key": "value"}
func(d)
```

```bash
$ uvx -w ty==0.0.21 ty check dict_str.py
Installed 1 package in 6ms
error[invalid-argument-type]: Argument to function `func` is incorrect
 --> dict_str.py:7:6
  |
6 | dstr = {"key": "value"}
7 | func(dstr)
  |      ^^^^ Expected `dict[str, str | None]`, found `dict[str, str]`
  |
info: Function defined here
 --> dict_str.py:1:5
  |
1 | def func(attrs: dict[str, str | None]) -> None:
  |     ^^^^ ---------------------------- Parameter declared here
2 |     pass
  |
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```

It was OK with version 0.0.20:

```shell
$ uvx -w ty==0.0.20 ty check dict_str.py
Installed 1 package in 4ms
All checks passed!
```

### Version

ty 0.0.21

**Comment by qmg-tmay** [comment]
See a similar issue also with ty 0.0.21:

```python
from collections.abc import Mapping
from typing import Any


def foo(rows: list[Mapping[str, Any]]) -> None:
    return


rows = [{"first": 1}, {"first": 2}]
foo(rows)
```

Expected `list[Mapping[str, Any]]`, found `list[dict[str, int]]`

**Comment by dhruvmanila** [comment]
This is an expected behavior because the value type of `dict` (which is mutable) is invariant which means that `str` is not assignable to `str | None` even though the former is a subtype of the latter. If you want the value types to be the same, then you can use `Mapping[str, str | None]` instead.

Considering your example, the type of `d` is inferred as `dict[str, str]` but given that `dict` is mutable the function call `func(d)` could mutate it and add a new entry where the value can be `None` given that you've typed the parameter `attrs` as such which means that the `d` after the function call can give us `None` which doesn't match the type of `d` before the function call. This error is to avoid this scenario.

**Comment by dhruvmanila** [comment]
@qmg-tmay the error you're seeing is for the same reason as I've mentioned above -- `list` is invariant on it's element type which means that you cannot assign `list[dict[...]]` to `list[Mapping[...]]` because the function can mutate `list`

**Comment by dhruvmanila** [comment]
I think we can improve the diagnostic message here given how Pyright provides additional explanation to this specific error case:
```
main.py:8:6: error: Argument of type "dict[str, str]" cannot be assigned to parameter "attrs" of type "dict[str, str | None]" in function "func"
  "dict[str, str]" is not assignable to "dict[str, str | None]"
    Type parameter "_VT@dict" is invariant, but "str" is not the same as "str | None"
    Consider switching from "dict" to "Mapping" which is covariant in the value type [reportArgumentType]
```

**Comment by AlexWaygood** [comment]
This error would also be avoided if we implemented https://github.com/astral-sh/ty/issues/1473, because we would use the context of the later use to inform the initial inference of the dictionary.

**Comment by qmg-tmay** [comment]
> [@qmg-tmay](https://github.com/qmg-tmay) the error you're seeing is for the same reason as I've mentioned above -- `list` is invariant on it's element type which means that you cannot assign `list[dict[...]]` to `list[Mapping[...]]` because the function can mutate `list`

Right I see. My assumption had been that this would work given that `dict` is a registered subclass of `Mapping` and `int` is a registered subclass of `Any`

**Comment by AlexWaygood** [comment]
In both cases, you can also avoid the error by giving the assignment an explicit annotation:

```py
def func(attrs: dict[str, str | None]) -> None:
    pass

func({"key": "value"})

d: dict[str, str | None] = {"key": "value"}
func(d)
```

and

```py
from collections.abc import Mapping
from typing import Any


def foo(rows: list[Mapping[str, Any]]) -> None:
    return


rows: list[Mapping[str, Any]] = [{"first": 1}, {"first": 2}]
foo(rows)
```

https://play.ty.dev/14e7d01f-e82c-4b0f-ac70-acfca1acf6cb

Providing an "expected type" in the annotation guides ty towards inferring the correct type for the assignment, which then avoids the type-mismatch error later on.

**Comment by bartfeenstra** [comment]
I **think** I ran into this as well, and if that is the case and I understand the relevant variance/mutability issues well enough, I believe I may a suggestion.

Given:
```python
from collections.abc import Mapping


class Locale:
    pass


translations: Mapping[Locale | str | None, str] | Mapping[Locale | None, str] = {
    "hello": "world",
}
```

ty says:
```
error[invalid-assignment]: Object of type `dict[str, str]` is not assignable to `Mapping[Locale | str | None, str] | Mapping[Locale | None, str]`
  --> test.py:8:15
   |
 8 |   translations: Mapping[Locale | str | None, str] | Mapping[Locale | None, str] = {
   |  _______________---------------------------------------------------------------___^
   | |               |
   | |               Declared type
 9 | |     "hello": "world",
10 | | }
   | |_^ Incompatible value of type `dict[str, str]`
   |
```

If I understand correctly, the usual reasons for the kind of dict incompatibility this thread relate to dictionaries being mutable, and that variance must be resolved more strictly because of that. In my code example above, I am wondering if those same rules apply, because the dictionary is never assigned to anything else, and as such can never be mutated in a way that is type-safe. Would it be possible for ty to relax 'variance strictness' in these cases, or are there reason I don't (yet) comprehend why that would not be possible?

Note that the union of mappings seems odd, but that is a simplification of existing nested union types in my codebase.

**Comment by ibraheemdev** [comment]
@bartfeenstra `Mapping` is invariant in its key type, but in your case we should be able to infer `dict[Local | str | None, str]` for the dictionary directly, the problem is we currently don't handle the union type annotation very well. This will be fixed by https://github.com/astral-sh/ruff/pull/23844.

**Comment by AlexWaygood** [comment]
@sharkdp is this now fixed by your recent work?

**Comment by sharkdp** [comment]
Yes, for the OP example, we now emit this diagnostic:
```
error[invalid-argument-type]: Argument to function `func` is incorrect
 --> main.py:8:6
  |
7 | d = {"key": "value"}
8 | func(d)
  |      ^ Expected `dict[str, str | None]`, found `dict[str, str]`
  |
info: Function defined here
 --> main.py:2:5
  |
1 | ## File: dict_str.py
2 | def func(attrs: dict[str, str | None]) -> None:
  |     ^^^^ ---------------------------- Parameter declared here
3 |     pass
  |
info: `dict` is invariant in its second type parameter
info: Consider using the supertype `collections.abc.Mapping`, which is covariant in its value type
info: For more information, see https://docs.astral.sh/ty/reference/typing-faq/#invariant-generics
info: rule `invalid-argument-type` is enabled by default

Found 1 diagnostic
```
